packages/kas/src/nodes.ts

```diff
@@ -949,153 +981,145 @@ export class Mul extends Seq {
+        const hasAdd = _.any(factors, isAdd);
+        const hasInverseAdd = _.any(factors, isInverseAdd);
 
         if (!(hasAdd || hasInverseAdd)) {
             return mul;
         }
 
-        var terms = _.groupBy(mul.terms, isInverse);
-        var normals = terms[false] || [];
-        var inverses = terms[true] || [];
+        let [inverses, normals] = partition(factors, isInverse);
 
         if (hasAdd) {
-            var grouped = _.groupBy(normals, isAdd);
-            var adds = grouped[true] || [];
-            var others = grouped[false] || [];
+            const [adds, others] = partition(normals, isAdd);
 
             // loop over each additive sequence
-            var expanded = _.reduce(
-                adds,
+            const expanded = _.reduce(
+                adds as Add[],
                 function (expanded, add) {
                     // loop over each expanded array of terms
                     return _.reduce(
                         expanded,
                         function (temp, array) {
                             // loop over each additive sequence's terms
                             return temp.concat(
-                                _.map(add.terms, function (term) {
-                                    return array.concat(term);
-                                }),
+                                _.map(add.terms, (term) => array.concat(term)),
                             );
                         },
-                        [],
+                        [] as Expr[][],
                     );
                 },
-                [[]],
+                [[]] as Expr[][],
             );
 
             // join each fully expanded array of factors with remaining multiplicative factors
-            var muls = _.map(expanded, function (array) {
+            const muls = _.map(expanded, function (array) {
                 return new Mul(others.concat(array)).flatten();
             });
 
             normals = [new Add(muls)];
         }
 
         if (hasInverseAdd) {
-            var denominator = new Mul(
+            const denominator = new Mul(
                 _.invoke(inverses, "getDenominator"),
             ).flatten();
-            inverses = [new Pow(denominator.expand(), Num.Div)];
+            inverses = [new Pow(denominator.expand(), NumDiv)];
         }
 
         return new Mul(normals.concat(inverses)).flatten();
     }
 
-    factor(options) {
+    factor(options): Expr {
         var factored = this.recurse("factor", options).flatten();
         if (!(factored instanceof Mul)) {
             return factored;
         }
 
         // Combine any factored out Rationals into one, but don't collect
-        var grouped = _.groupBy(factored.terms, function (term) {
-            return term instanceof Rational;
+        var grouped = _.groupBy(factored.terms, (term) => {
+            return term instanceof Rational ? "true" : "false";
```

Comment:
Why "true"/"false" vs true/false?

EDIT: oh I see, `groupBy` is using the strings as keys in a return object. I think this is confusing, but your solution (`grouped.true`) is less confusing than the original (`grouped[true]`). I'm just wondering if we should have true/"true"/false/"false" as property names at all.

Reply:
I need to update this to use the `partition` helper function I added.